src/inc/til/small_vector.h

```diff
@@ -807,51 +810,69 @@ namespace til
+            if constexpr (noexcept(func(begin())) && std::is_trivially_destructible_v<T> && std::is_trivially_copyable_v<T>)
+            {
+                _size = new_size;
 
-            // 1. We've resized the vector to fit an additional `count` items.
-            //    Initialize the `count` items at the end by moving existing items over.
-            //
-            // This line is noexcept:
-            // It can't throw because of the earlier static_assert(std::is_nothrow_move_assignable_v<T>).
-            std::uninitialized_move(end() - displacement, end(), _uninitialized_begin() + (new_size - displacement));
-            _size = new_size;
+                // Make space for `count` uninitialized items, to be filled by `func`.
+                const auto it = begin() + offset;
+                const auto dest = it + count;
+                memmove(dest._Unwrapped(), it._Unwrapped(), moveable * sizeof(T));
 
-            // 2. Now that all `new_size` items are initialized we can move as many
-            //    items towards the end as we need to make space for `count` items.
-            //
-            // This line is noexcept:
-            // It can't throw because of the earlier static_assert(std::is_nothrow_move_assignable_v<T>).
-            const auto displacement_beg = begin() + offset;
-            const auto displacement_end = displacement_beg + displacement;
-            std::move_backward(displacement_beg, displacement_beg + (moveable - displacement), end() - displacement);
-
-            // 3. Destroy the displaced existing items, so that we can use
-            //    std::uninitialized_*/std::construct_* functions inside `func`.
-            //    This isn't optimal, but better than nothing.
-            std::destroy(displacement_beg, displacement_end);
-
-            try
-            {
-                func(displacement_beg);
+                func(it);
+                return it;
             }
-            catch (...)
+            else
             {
-                std::destroy(displacement_end, end());
-                _size = offset;
-                throw;
-            }
+                // The earlier static_assert(std::is_nothrow_move_assignable_v<T>)
+                // ensures that we don't exit in a weird state with invalid `_size`.
 
-            return displacement_beg;
+                // 1. We've resized the vector to fit an additional `count` items.
+                //    Initialize the `count` items at the end by moving existing items over.
+                //
+                // This line is noexcept:
+                // It can't throw because of the earlier static_assert(std::is_nothrow_move_assignable_v<T>).
+                const auto displacement = std::min(count, moveable);
+                std::uninitialized_move(end() - displacement, end(), _uninitialized_begin() + (new_size - displacement));

```

Comment:
do we have tests exercising this code? validating it?

Reply:
Yes, parts of the Basic tests cover this.